modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/ConditionalProcessor.java

```diff
@@ -0,0 +1,439 @@
+        this.scriptService = scriptService;
+        this.processor = processor;
+    }
+
+    @Override
+    public void execute(IngestDocument ingestDocument) throws Exception {
+        if (scriptService.compile(condition, ProcessorConditionalScript.CONTEXT)
+            .newInstance(condition.getParams()).execute(new UnmodifiableIngestData(ingestDocument.getSourceAndMetadata()))) {
+            processor.execute(ingestDocument);
+        }
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    public static final class Factory implements Processor.Factory {
+
+        private final ScriptService scriptService;
+
+        public Factory(ScriptService scriptService) {
+            this.scriptService = scriptService;
+        }
+
+        @Override
+        public ConditionalProcessor create(Map<String, Processor.Factory> factories, String tag,
+            Map<String, Object> config) throws Exception {
+            Map<String, Map<String, Object>> processorConfig = readMap(TYPE, tag, config, "processor");
+            final Script script;
+            try (XContentBuilder builder = XContentBuilder.builder(JsonXContent.jsonXContent)
+                .map(normalizeScript(config.get("script")));
+                 InputStream stream = BytesReference.bytes(builder).streamInput();
+                 XContentParser parser = XContentType.JSON.xContent().createParser(NamedXContentRegistry.EMPTY,
+                     LoggingDeprecationHandler.INSTANCE, stream)) {
+                script = Script.parse(parser);
+                config.remove("script");
+                // verify script is able to be compiled before successfully creating processor.
+                try {
+                    scriptService.compile(script, ProcessorConditionalScript.CONTEXT);
+                } catch (ScriptException e) {
+                    throw newConfigurationException(TYPE, tag, null, e);
+                }
+            }
+            Map.Entry<String, Map<String, Object>> entry = processorConfig.entrySet().iterator().next();
+            Processor processor = ConfigurationUtils.readProcessor(factories, entry.getKey(), entry.getValue());
+            return new ConditionalProcessor(tag, script, scriptService, processor);
+        }
+
+        @SuppressWarnings("unchecked")
+        private static Map<String, Object> normalizeScript(Object scriptConfig) {
+            if (scriptConfig instanceof Map<?, ?>) {
+                return (Map<String, Object>) scriptConfig;
+            } else if (scriptConfig instanceof String) {
+                return Collections.singletonMap("source", scriptConfig);
+            } else {
+                throw newConfigurationException(TYPE, null, "script",
+                    "property isn't a map or string, but of type [" + scriptConfig.getClass().getName() + "]");
+            }
+        }
+    }
+
+    private static Object wrapUnmodifiable(Object raw) {
+        if (raw instanceof Map) {
```

Comment:
Can you add a comment here that these types must match what the json parser can create, and that anything not handled here must be immutable already (eg boxed numerics)?

Reply:
sure :)